### PR TITLE
First DDEve config for FCCSW.

### PR DESCRIPTION
**Pull request by vvolkl**
New subfolder containing documentation and a basic configuration file for use with DDEve (see documentation). This is WIP, but mainly on the DD4hep side -- depending on the required features, the configuration should be all that is needed in FCCSW.

Update: the necessary changes in dd4hep were included with r2299 (see https://sft.its.cern.ch/jira/browse/DDFORHEP-23). The FCC installation of DD4hep will be updated soon.

### CODE REVIEW

**Comment by jlingema** [comment]
Hi Valentin - this is great. Is there a way to automatically test this (i.e. the configuration)?

**Comment by vvolkl** [comment]
Good point, the full functionality definitely not, but the syntax of the xml and the detector names probably yes. I have an open ticket with DD4hep regarding the cli, I will bring testing up there.

On 16 Aug 2016, at 06:03, Joschka Lingemann <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi Valentin - this is great. Is there a way to automatically test this (i.e. the configuration)?

File: Visualization/src/FCCEventHandler.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "podio/EventStore.h"
+#include "podio/ROOTReader.h"
+
+#include <map>
+#include <string>
+
+namespace DD4hep {
+
+class FCCEventHandler : public EventHandler {
```
Comment (hegner) [inline Visualization/src/FCCEventHandler.h]: why is the FCCEventHandler in the DD4hep namespace?

File: Visualization/src/FCCEventHandler.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+protected:
```
Comment (hegner) [inline Visualization/src/FCCEventHandler.h]: make it private. I don't think we want something to inherit from this class, no?

File: Visualization/src/FCCEventHandler.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+public:
+  /// Standard constructor
+  FCCEventHandler();
+  /// Default destructor
+  virtual ~FCCEventHandler();
+
+  // virtual bool hasEvent() { std::cout << "has Event" <<std::endl; return true;};
+  /// Access the map of simulation data collections
+  virtual const TypedEventCollections& data() const { return m_data; }
```
Comment (hegner) [inline Visualization/src/FCCEventHandler.h]: a good practice is using the `override` keyword here.

File: Visualization/src/FCCEventHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+long FCCEventHandler::numEvents() const {
+  if (m_hasFile) {
+    return m_ROOTReader.getEntries();
+  }
+  return -1;
+}
+
+/// Access to the collection type by name
+EventHandler::CollectionType FCCEventHandler::collectionType(const std::string& cl) const {
+  if (cl == "caloClusters")
```
Comment (hegner) [inline Visualization/src/FCCEventHandler.cpp]: Eventually we should put that into some external config.

File: Visualization/src/FCCEventHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    return PARTICLE_COLLECTION;
+  else
+    return NO_COLLECTION;
+}
+
+
+
+/// Call functor on hit collection
+size_t FCCEventHandler::collectionLoop(const std::string& collection, DDEveHitActor& actor) {
+  /// TODO: avoid code duplication
```
Comment (hegner) [inline Visualization/src/FCCEventHandler.cpp]: could you remove the duplication before merging?

Comment (vvolkl) [inline Visualization/src/FCCEventHandler.cpp]: I wasn't  sure how to do this without templating, but  I'll take another look!

Comment (hegner) [inline Visualization/src/FCCEventHandler.cpp]: templating is the most straight forward approach.

File: Visualization/compact/DDEve_FCCSW.xml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    <detelement name="GenericTrackerBarrel" load_geo="-1" alpha="0.5"/>
+    <calodata name="trackerdata" />
+    <calodata name="calohits" />
+  </view>
+  <view name="test_multiview" type="MultiView">
+    <detelement name="GenericTrackerBarrel" load_geo="-1" alpha="0.5"/>
+    <calodata name="trackerdata" />
+  </view>
+  <include ref="/home/vavolkl/FCC/FCCSW/Detector/DetFCChhBaseline1/compact/FCChh_DectEmptyMaster.xml"/>
+  <include ref="/home/vavolkl/FCC/FCCSW/Detector/DetFCChhTrackerSimple/compact/Tracker.xml"/>
```
Comment (zaborowska) [inline Visualization/compact/DDEve_FCCSW.xml]: can you remove first part of path?

Comment (vvolkl) [inline Visualization/compact/DDEve_FCCSW.xml]: oh, that slipped throught the cracks. thanks for noticing!

**Comment by vvolkl** [comment]
There is one more fix needed in DD4hep, otherwise the rootmaps do not work, but with a custom installation of dd4hep this should already work.

**Comment by jlingema** [comment]
I've based #134 on this one and migrated the DD4hep converter to the new EDM, so I'll close this one.

**Pull request closed without merging**